test/Fencer/Logic/Test.hs

```diff
@@ -0,0 +1,103 @@
+{-# LANGUAGE OverloadedLabels  #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes       #-}
+
+-- | Tests for "Fencer.Logic".
+module Fencer.Logic.Test
+  ( test_logicLimitUnitChange
+  ) where
+
+import           BasePrelude
+
+import           Data.List.NonEmpty (NonEmpty((:|)))
+import qualified Data.List.NonEmpty as NE
+import           Test.Tasty (TestTree)
+import           Test.Tasty.HUnit (assertBool, testCase)
+
+import           Fencer.Counter (CounterStatus, counterRemainingLimit)
+import           Fencer.Logic (AppState, setRules, updateLimitCounter)
+import           Fencer.Rules (definitionsToRuleTree)
+import           Fencer.Server.Test (withServer, serverAppState)
+import           Fencer.Types
+
+
+-- | Test that a rule limit unit change adds a new counter and leaves
+-- the old one intact.
+test_logicLimitUnitChange :: TestTree
+test_logicLimitUnitChange =
+  withServer $ \serverIO ->
+    testCase "A rule limit unit change on rule reloading" $ do
+      state <- serverAppState <$> serverIO
+      void $ atomically $ setRules state (mapRuleDefs definitions1)
+
+      -- Record a hit and get a remaining limit
+      st1 <- getRemainingLimit <$> makeAHit state
+      assertBool
+        "The remaining rate limit was not updated!"
+        (st1 == limit - hits)
+
+      -- Set the new rules and the rules reloaded flag
+      atomically $ setRules state (mapRuleDefs definitions2)
+      -- Record a hit and get a remaining limit
+      st2 <- getRemainingLimit <$> makeAHit state
+      assertBool
```

Comment:
You should use `assertEqual` to get better errors from tasty when the test fails.

I pushed a commit that adds an HLint rule to this effect, you can do `hlint --git` to see it in action.

Reply:
A good point! Somehow this slipped through. Took care of it in an upcoming commit.

Haven't checked, but it'd be great to have `hlint` run as a Git hook or something similar.

Reply:
You can add a hook locally + add instructions for how to do so to CONTRIBUTING.md. 

HLint also runs on Code Climate. It used to ignore `tests/` but starting from the next commit it should work on `tests/` too.

Reply:
> https://codeclimate.com/github/juspay/fencer/pull/69

Fixed.